testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java

```diff
@@ -0,0 +1,907 @@
+    verify(mockClientStreamListener, never()).closed(any(Status.class), any(Metadata.class));
+  }
+
+  @Test
+  public void serverCancel() throws Exception {
+    server.start(serverListener);
+    client.start(mockClientTransportListener);
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(1000, TimeUnit.MILLISECONDS);
+    serverTransport = serverTransportListener.transport;
+
+    ClientStream clientStream = client.newStream(methodDescriptor, new Metadata());
+    clientStream.start(mockClientStreamListener);
+    StreamCreation serverStreamCreation
+        = serverTransportListener.takeStreamOrFail(1000, TimeUnit.MILLISECONDS);
+    ServerStream serverStream = serverStreamCreation.stream;
+    ServerStreamListener mockServerStreamListener = serverStreamCreation.listener;
+
+    Status status = Status.DEADLINE_EXCEEDED.withDescription("It was bound to happen");
+    serverStream.cancel(status);
+    verify(mockServerStreamListener, timeout(1000)).closed(Matchers.same(status));
+    verify(mockClientStreamListener, timeout(1000))
+        .closed(statusCaptor.capture(), any(Metadata.class));
+    // Presently we can't sent much back to the client in this case. Verify that is the current
+    // behavior for consistency between transports.
+    assertCodeEquals(Status.CANCELLED, statusCaptor.getValue());
+
+    // Second cancellation shouldn't trigger additional callbacks
+    reset(mockServerStreamListener);
+    reset(mockClientStreamListener);
+    serverStream.cancel(status);
+    verify(mockServerStreamListener, never()).closed(any(Status.class));
+    verify(mockClientStreamListener, never()).closed(any(Status.class), any(Metadata.class));
+  }
+
+  @Test
+  public void flowControlPushBack() throws Exception {
+    server.start(serverListener);
+    client.start(mockClientTransportListener);
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(1000, TimeUnit.MILLISECONDS);
+    serverTransport = serverTransportListener.transport;
+
+    ClientStream clientStream = client.newStream(methodDescriptor, new Metadata());
+    clientStream.start(mockClientStreamListener);
+    StreamCreation serverStreamCreation
+        = serverTransportListener.takeStreamOrFail(1000, TimeUnit.MILLISECONDS);
+    assertEquals(methodDescriptor.getFullMethodName(), serverStreamCreation.method);
+    ServerStream serverStream = serverStreamCreation.stream;
+    ServerStreamListener mockServerStreamListener = serverStreamCreation.listener;
+    serverStream.writeHeaders(new Metadata());
+
+    Answer<Void> closeStream = new Answer<Void>() {
+      @Override
+      public Void answer(InvocationOnMock invocation) throws Exception {
+        Object[] args = invocation.getArguments();
+        ((InputStream) args[0]).close();
+        return null;
+      }
+    };
+
+    String largeMessage;
+    {
```

Comment:
Readability. The local variables are thrown away and the side-effects of the code are limited. It allows you to easily ignore the block.